[inductor][bucketing] Fx collectives bucketing of multiple dtypes

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #162470

Bucketing of multiple dtypes to be processed in one bucketed collective.

First target is to bucket bf16 and f32, but already can be used with other dtypes.

For now multidtype bucketing is only supported with "custom_ops" mode.
Non custom_ops needs additional work on inductor side.



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @ezyang

IvanKobzarev committed:
[inductor][bucketing][WIP] Bucket multidtype

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor][bucketing][WIP] Bucket multidtype"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor][bucketing][WIP] Bucket multidtype"


Lowering for aten._to_copy fails on fallback and 
```
  File "/data/users/ivankobzarev/h/pytorch/torch/_dynamo/eval_frame.py", line 845, in compile_wrapper
    raise e.remove_dynamo_frames() from None  # see TORCHDYNAMO_VERBOSE=1
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/compile_fx.py", line 990, in _compile_fx_inner
    raise InductorError(e, currentframe()).with_traceback(
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/compile_fx.py", line 974, in _compile_fx_inner
    mb_compiled_graph = fx_codegen_and_compile(
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/compile_fx.py", line 1695, in fx_codegen_and_compile
    return scheme.codegen_and_compile(gm, example_inputs, inputs_to_check, graph_kwargs)
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/compile_fx.py", line 1420, in codegen_and_compile
    graph.run(*example_inputs)
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/graph.py", line 937, in run
    return super().run(*args)
  File "/data/users/ivankobzarev/h/pytorch/torch/fx/interpreter.py", line 174, in run
    self.env[node] = self.run_node(node)
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/graph.py", line 1624, in run_node
    result = super().run_node(n)
  File "/data/users/ivankobzarev/h/pytorch/torch/fx/interpreter.py", line 256, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/graph.py", line 1233, in call_function
    make_fallback(target, layout_constraint=decided_constraint)
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/lowering.py", line 2080, in make_fallback
    assert op not in decompositions or override_decomp, (
torch._inductor.exc.InductorError: AssertionError: both a fallback and a decomp for same op: aten._to_copy.default
```


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor][bucketing][WIP] Bucket multidtype"


Lowering for aten._to_copy fails on fallback and 
```
  File "/data/users/ivankobzarev/h/pytorch/torch/_dynamo/eval_frame.py", line 845, in compile_wrapper
    raise e.remove_dynamo_frames() from None  # see TORCHDYNAMO_VERBOSE=1
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/compile_fx.py", line 990, in _compile_fx_inner
    raise InductorError(e, currentframe()).with_traceback(
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/compile_fx.py", line 974, in _compile_fx_inner
    mb_compiled_graph = fx_codegen_and_compile(
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/compile_fx.py", line 1695, in fx_codegen_and_compile
    return scheme.codegen_and_compile(gm, example_inputs, inputs_to_check, graph_kwargs)
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/compile_fx.py", line 1420, in codegen_and_compile
    graph.run(*example_inputs)
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/graph.py", line 937, in run
    return super().run(*args)
  File "/data/users/ivankobzarev/h/pytorch/torch/fx/interpreter.py", line 174, in run
    self.env[node] = self.run_node(node)
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/graph.py", line 1624, in run_node
    result = super().run_node(n)
  File "/data/users/ivankobzarev/h/pytorch/torch/fx/interpreter.py", line 256, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/graph.py", line 1233, in call_function
    make_fallback(target, layout_constraint=decided_constraint)
  File "/data/users/ivankobzarev/h/pytorch/torch/_inductor/lowering.py", line 2080, in make_fallback
    assert op not in decompositions or override_decomp, (
torch._inductor.exc.InductorError: AssertionError: both a fallback and a decomp for same op: aten._to_copy.default
```


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

eellison reviewed (commented):
now that the other prs have landed - mind rebasing ?

IvanKobzarev committed:
Update on "[inductor][bucketing] Fx collectives bucketing of multiple dtypes"





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor][bucketing] Fx collectives bucketing of multiple dtypes"


Bucketing of multiple dtypes to be processed in one bucketed collective.

First target is to bucket bf16 and f32, but already can be used with other dtypes.

For now multidtype bucketing is only supported with "custom_ops" mode.
Non custom_ops needs additional work on inductor side.



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor][bucketing] Fx collectives bucketing of multiple dtypes"


Bucketing of multiple dtypes to be processed in one bucketed collective.

First target is to bucket bf16 and f32, but already can be used with other dtypes.

For now multidtype bucketing is only supported with "custom_ops" mode.
Non custom_ops needs additional work on inductor side.



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

eellison reviewed (commented):
Looks good ! 

Only blocking question is about reduce scatter multi dtype

IvanKobzarev commented on torch/_inductor/fx_passes/bucketing.py:
```diff
@@ -288,35 +324,52 @@ def bucket_reduce_scatter_by_mb(
     rs_ins: list[torch.Tensor],
     group_size: int,
+    dtype: torch.dtype,  # type: ignore[name-defined]
+    numel_mults: list[int],
```
Custom ops do not support List[dtype] to calculate it in the op. So I changed to passing multipliers as List[int] instead. We need to know how to split the result according to different dtypes.

IvanKobzarev commented on test/distributed/test_inductor_collectives.py:
```diff
@@ -1754,6 +1754,115 @@ def func2(x, w, rs_0, rs_1, tag, ranks, group_size):
+
+    @unittest.skipIf(not HAS_GPU, "Inductor+gpu needs triton and recent GPU arch")
+    @unittest.skipIf(not SM80OrLater, "bfloat16")
+    @parametrize("bucket_mode", ["all_custom_ops_multidtype"])
```
We can. It will be a bit long sequence, as we have 2**num_options (custom_ops +/ multidtype +/- all/fsdp). We should reduce it to 4 for now.

IvanKobzarev committed:
Update on "[inductor][bucketing] Fx collectives bucketing of multiple dtypes"


Bucketing of multiple dtypes to be processed in one bucketed collective.

First target is to bucket bf16 and f32, but already can be used with other dtypes.

For now multidtype bucketing is only supported with "custom_ops" mode.
Non custom_ops needs additional work on inductor side.



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor][bucketing] Fx collectives bucketing of multiple dtypes"


Bucketing of multiple dtypes to be processed in one bucketed collective.

First target is to bucket bf16 and f32, but already can be used with other dtypes.

For now multidtype bucketing is only supported with "custom_ops" mode.
Non custom_ops needs additional work on inductor side.



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

eellison reviewed (approved):
looks good, just one comment

IvanKobzarev committed:
Update on "[inductor][bucketing] Fx collectives bucketing of multiple dtypes"


Bucketing of multiple dtypes to be processed in one bucketed collective.

First target is to bucket bf16 and f32, but already can be used with other dtypes.

For now multidtype bucketing is only supported with "custom_ops" mode.
Non custom_ops needs additional work on inductor side.



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor][bucketing] Fx collectives bucketing of multiple dtypes"


Bucketing of multiple dtypes to be processed in one bucketed collective.

First target is to bucket bf16 and f32, but already can be used with other dtypes.

For now multidtype bucketing is only supported with "custom_ops" mode.
Non custom_ops needs additional work on inductor side.



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor][bucketing] Fx collectives bucketing of multiple dtypes"


Bucketing of multiple dtypes to be processed in one bucketed collective.

First target is to bucket bf16 and f32, but already can be used with other dtypes.

For now multidtype bucketing is only supported with "custom_ops" mode.
Non custom_ops needs additional work on inductor side.



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]
